### PR TITLE
Fix building for host architecture by default

### PR DESCRIPTION
**Pull request by RoboTux**
By default Tensorflow build aims to only build what is relevant for the host targets. However, due to some missing guards here and there code for other architecture may end up being built. For instance, when building tensorflow/compiler/mlir:tf-opt on AArch64 code specific to x86 also ends up being built. This commit fixes that (aka issue #59595).

### CODE REVIEW

**Comment by cheshire** [comment]
@tpopp could you take a look?

File: tensorflow/compiler/xla/runtime/BUILD
Comment (tpopp) [inline tensorflow/compiler/xla/runtime/BUILD:268]: Is it purposeful that some of these dependencies were fully removed?

**Review by tpopp — DISMISSED** [review]
Thanks for doing this. It is not feasible for me to understand if all required dependencies are supported with this change, so I'll run the test suite and base any feedback on failing CI or all passing CI.

File: tensorflow/compiler/xla/runtime/BUILD
Comment (RoboTux) [reply tensorflow/compiler/xla/runtime/BUILD:268]: No, I must have made a mistake in my rebase. Let me correct that.

**Comment by tpopp** [comment]
Checking, should I review this again or not (it is marked as draft)?

**Comment by RoboTux** [comment]
> Hi @tpopp! @RoboTux is ooo until April 11th. He asked me to let you know he wanted to have a look at the CI failures before taking it out of draft. Cheers and have a nice weekend!

I'm slowly catching up but the problem seems to be that tensorflow/python/tools/saved_model_cli_test.py assumes both x86 and aarch64 are always available. I'll see how to remove that expectation on Monday, any suggestion welcome though.

**Comment by tpopp** [comment]
I'll share it with others in the US in hopes that someone else has thoughts but won't be able to look earlier than Monday or Tuesday myself.

**Comment by RoboTux** [comment]
FYI: I have a revised patch that I'm going to push shortly (probably tomorrow).

**Comment by tpopp** [comment]
Hey, sorry, I just came back from vacation. I'm going to see if I can find a more appropriate person to review this and will try to, in parallel, make sure everything passes, so there are no unexpected surprises later.

**Comment by ddunl** [comment]
To me it seems like @jakeharmon8 and @MichaelHudgins might be the ones who should look at this?

**Comment by RoboTux** [comment]
Failure seems unrelated:
"ModuleNotFoundError: No module named 'tensorflow.compiler.mlir.stablehlo'"

It also passed CI before this last merge.

**Comment by RoboTux** [comment]
> Failure seems unrelated: "ModuleNotFoundError: No module named 'tensorflow.compiler.mlir.stablehlo'"
> 
> It also passed CI before this last merge.

Same error on master as well: https://github.com/tensorflow/tensorflow/actions/runs/5010911951/jobs/8981212479

**Comment by elfringham** [comment]
@RoboTux That failure was introduced with https://github.com/tensorflow/tensorflow/commit/c09dcc2c7d07a84cdb9f9b5342d7803f5dbc50b3

**Comment by MichaelHudgins** [comment]
@vam-google or @jakeharmon8 could you PTAL?

**Comment by RoboTux** [comment]
Ping?

**Review by jakeharmon8 — APPROVED** [review]

**Comment by RoboTux** [comment]
I'm working on fixing remaining issues.

**Comment by RoboTux** [comment]
The latest push should fix the API and codestyle issues. Note: this does indeed introduce a new function in the test API and follows a similar pattern to ROCm and cuda. Let me know if it isn't fine.

**Comment by RoboTux** [comment]
All remaining issues seem unrelated to this change:
- Arm CI fails to find the pip package it created (ls: cannot access '/tensorflow/whl/-*310*310*.whl': No such file or directory)
- AMD ROCm fails to find clang
- MacOS has 16 tests timing out

**Comment by RoboTux** [comment]
Ping?

**Comment by MichaelHudgins** [comment]
@jakeharmon8 can you rereview? 
@RoboTux can you please rebase?

**Review by jakeharmon8 — APPROVED** [review]

**Comment by RoboTux** [comment]
Windows build failure looks unrelated:

```
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(118): error C2065: 'not': undeclared identifier
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(118): error C2146: syntax error: missing ')' before identifier 'scalables'
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(118): error C2059: syntax error: ')'
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(119): error C2146: syntax error: missing ';' before identifier 'printer'
```

**Comment by RoboTux** [comment]
MacOS build seems to fail for performance reasons if I read this correctly:
```
//tensorflow/python/kernel_tests/linalg/sparse:csr_sparse_matrix_sparse_mat_mul_grad_test_cpu FAILED in 150 out of 150 in 1.2s
  Stats over 150 runs: max = 1.2s, min = 0.5s, avg = 0.8s, dev = 0.1s
```

Since this code only changes the build system that must be unrelated.

As to ARM CI there is one test where TF crashes which I'm not sure about and one x86 dialect loading failure which is most likely related to this diff. I'll have a look now

```
FAIL: //tensorflow/compiler/xla/mlir/math/transforms/tests:math_optimization.mlir.test
(...)
LLVM ERROR: Building op `x86vector.avx.rsqrt` but it isn't registered in this MLIRContext: the dialect may not be loaded or this operation isn't registered by the dialect.
```

**Comment by elfringham** [comment]
//tensorflow/dtensor/mlir/tests:spmd_expansion.mlir.test is known as flaky, see https://github.com/tensorflow/tensorflow/issues/61164

**Review by jakeharmon8 — APPROVED** [review]

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

**Comment by freddan80** [comment]
fyi - @RoboTux is on leave until 26th.

**Comment by johnnkp** [comment]
I can confirm that this PR works for Windows CUDA build. The remaining are resolving conflicts and merge quickly before new merge conflicts happens.

**Review by jakeharmon8 — APPROVED** [review]

File: tensorflow/tsl/platform/default/build_config.bzl
Comment (hawkinsp) [inline tensorflow/tsl/platform/default/build_config.bzl:873]: Aren't we missing Mac ARM here (and Mac x86 below as well)?

Comment (RoboTux) [reply tensorflow/tsl/platform/default/build_config.bzl:873]: Indeed, missed that sorry. I'll upload a new version.

**Comment by RoboTux** [comment]
Ping?

**Review by jakeharmon8 — APPROVED** [review]

File: tensorflow/python/platform/test.py
Comment (jakeharmon8) [inline tensorflow/python/platform/test.py:201]: Add an "Args" section here to appease our linter (e.g. see line 82)

**Review by jakeharmon8 — CHANGES_REQUESTED** [review]
Add a note about the API addition in RELEASE.md (an internal check requires it when making API changes)

**Comment by jakeharmon8** [comment]
There are two more blockers that I'm working on:

1) Internal technical issues that I should have fixed soon
2) Changes to the API normally require an RFC+Review (see [here](https://github.com/tensorflow/community/blob/master/rfcs/yyyymmdd-rfc-template.md)). I'm looking into what the process is for external contributors, and whether or not there's a shorter process for small changes like this. If you'd like to avoid that red tape, you could refactor this to remove the API changes.

**Comment by RoboTux** [comment]
> There are two more blockers that I'm working on:
> 
>     1. Internal technical issues that I should have fixed soon
> 
>     2. Changes to the API normally require an RFC+Review (see [here](https://github.com/tensorflow/community/blob/master/rfcs/yyyymmdd-rfc-template.md)). I'm looking into what the process is for external contributors, and whether or not there's a shorter process for small changes like this. If you'd like to avoid that red tape, you could refactor this to remove the API changes.

I'm not sure how I could remove the API change. I need a way to skip a target CPU specific test if the support for that target is not built in so need a way to query that.

**Review by jakeharmon8 — APPROVED** [review]

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

**Review by jakeharmon8 — APPROVED** [review]

**Review by jakeharmon8 — APPROVED** [review]

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

**Comment by jakeharmon8** [comment]
The conflicts are that many of these files moved - I think I can just merge the internal version (since they didn't move internally). Let me try getting all the approvals again tomorrow and hopefully I can ship this.

**Comment by jakeharmon8** [comment]
When these files moved, copybara deleted the internal CL and created a new one, which means I have to start over on all of the internal approvals. Sorry for the comical amount of red tape - I should have this merged any day now.

**Comment by ebrevdo** [comment]
Sorry for the late join on this PR.  The current behavior of XLA AOT generation is to support x-compiling across multiple architectures.  This is used, by example, by the `saved_model_cli` tool.  It's OK to disable when building TF in general, but for the PyPi build we should allow the existing behavior.  Can you ensure that this PR continues to build TF PyPi packages that allow XLA AOT cross-compilation across all existing known architectures?

**Comment by RoboTux** [comment]
> Sorry for the late join on this PR. The current behavior of XLA AOT generation is to support x-compiling across multiple architectures. This is used, by example, by the `saved_model_cli` tool. It's OK to disable when building TF in general, but for the PyPi build we should allow the existing behavior. Can you ensure that this PR continues to build TF PyPi packages that allow XLA AOT cross-compilation across all existing known architectures?
> 
> I _think_ the changes would have to be made in the files in: ci/official/envs

I'll double check.

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

File: ci/official/envs/continuous_linux_x86_cpu_py310
Comment (MichaelHudgins) [inline ci/official/envs/continuous_linux_x86_cpu_py310:1]: You can update in only a single place within the root .bazelrc instead of in each env if this is needed in either all release configs or all linux release configs by adding to release_base or release_cpu_linux respectively.

**Review by MichaelHudgins — CHANGES_REQUESTED** [review]

File: ci/official/envs/continuous_linux_x86_cpu_py310
Comment (MichaelHudgins) [reply ci/official/envs/continuous_linux_x86_cpu_py310:1]: Also side note, those envs are currently not used by the prod jobs.  We will be switching to them in the near future though.  Adding to the root .bazelrc will catch both the older and newer ci jobs.

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

**Comment by ebrevdo** [comment]
We should get this in!

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

**Comment by RoboTux** [comment]
> Hi @RoboTux Can you please resolve conflicts? Thank you!

Done.

Note 1: I'm likely to soon have less time to push this patch forward and my have to drop it if it doesn't get approved soon.
Note 2: The patch contains changes to third_party/xla and third_party/xla/third_party/tsl. I'm looking for a general approval before submitting these changes separately to the appropriate repos.

Best regards,
Thomas

**Comment by RoboTux** [comment]
Ping @jakeharmon8 ?

**Review by jakeharmon8 — APPROVED** [review]

**Comment by RoboTux** [comment]
The PyLint issue is unrelated. It is due to the import of gen_sparse_ops added by https://github.com/tensorflow/tensorflow/commit/d7030ffa0e0a90378f3a4efc099cc50fd07c0577

**Comment by jakeharmon8** [comment]
The Kafkaesque nightmare continues - two internal approvals left.

**Comment by RoboTux** [comment]
> The Kafkaesque nightmare continues - two internal approvals left.

Thanks for the update and for pushing this. Much appreciated.

**Comment by gbaned** [comment]
Hi @RoboTux Can you please resolve conflicts? Thank you!

**Comment by jakeharmon8** [comment]
Don't worry about rebasing - I'm still managing this internally. It had broken some internal tests concerning LLVM and Arm64, but I think I've almost got that working. I'll jump back on this today.

**Comment by ddunl** [comment]
It seems like this landed, so we can close. @jakeharmon8 ?

**Comment by RoboTux** [comment]
> It seems like this landed, so we can close. @jakeharmon8 ?

If it indeed has thanks Jake!

**Comment by jakeharmon8** [comment]
Yup, it finally landed. Sorry for the delay - this PR really was the perfect storm of creating as much internal bureaucracy as possible!

**Pull request closed without merging**
